### PR TITLE
WAITP-1431: Update datatrak survey url with country code

### PR DESCRIPTION
**Pull request by tcaiger**
### Changes:

- Make TupaiaWebEntityRequest a re-usable type
- Create a new SingleEntityRoute
- Add countryCode to survey routes & refactor from user preference country.code to urlParams countryCode

### CODE REVIEW

**Review by acdunham — CHANGES_REQUESTED** [review]
This is looking really good! Just noticed that the recent surveys and recent responses tiles don't direct to have the country in the URL though, which causes an error. 

I guess it might also open up the question - should project code be in the URL as well? I had a scenario, for example, where I went directly to a survey URL for a survey in Tonga, but my previously selected project was STRIVE, which meant I got an error fetching entities because STRIVE project doesn't have Tonga as a country. So we would get similar issues to what we had before the country was in the URL. What do you think?

**Comment by tcaiger** [comment]
> This is looking really good! Just noticed that the recent surveys and recent responses tiles don't direct to have the country in the URL though, which causes an error.
> 
> I guess it might also open up the question - should project code be in the URL as well? I had a scenario, for example, where I went directly to a survey URL for a survey in Tonga, but my previously selected project was STRIVE, which meant I got an error fetching entities because STRIVE project doesn't have Tonga as a country. So we would get similar issues to what we had before the country was in the URL. What do you think?

Oh yes good questions. Here's a summary from our chat.

- Don’t need project code in the url as we can use survey.project_id
- Auto update user preferred country and project on deep linking
- Add toasts whenever user preferences are changed

File: packages/datatrak-web-server/src/routes/SubmitSurvey/processSurveyResponse.ts [hunk 1]
```diff
@@ -51,7 +51,6 @@ export const processSurveyResponse = async (
     survey_id: surveyId,
     start_time: startTime,
     entity_id: countryId,
-    country_id: countryId,
```
Comment (tcaiger) [inline packages/datatrak-web-server/src/routes/SubmitSurvey/processSurveyResponse.ts:54]: This isn't required in the central server

File: packages/datatrak-web/src/views/LandingPage/RecentSurveysSection.tsx [hunk 2]
```diff
@@ -44,25 +42,14 @@ const ScrollBody = styled.div`
 
 export const RecentSurveysSection = () => {
   const { data: recentSurveys = [], isSuccess, isLoading } = useCurrentUserRecentSurveys();
-  const navigate = useNavigate();
-  const { mutateAsync: editUser } = useEditUser();
-
-  const handleSelectSurvey = async (surveyCode: string, countryId: string) => {
```
Comment (tcaiger) [inline packages/datatrak-web/src/views/LandingPage/RecentSurveysSection.tsx:50]: This is no longer needed as the countryId is checked and updated in the survey page now

**Review by acdunham — APPROVED** [review]
Thanks for those changes!

**Pull request merged**